### PR TITLE
Make compiler dead code warning spans use name instead of entire block

### PR DESCRIPTION
**Pull request by sdankel**
Closes https://github.com/FuelLabs/sway/issues/3746

This PR changes the highlighted portion for dead code warnings to show only the declaration name (ident) rather than the entire code block for the LSP.

It also adds some test helper functions:
- an async function that spawns a thread, uses the server socket to capture server-initiated requests, and asserts that the messages received match the expected messages
- a simple function to read a json file into memory to be used as a test fixture

And adds a dead code warning for Enum, allowing us to highlight only the enum name, rather than the entire block.

Before:

image

After:

image

### CODE REVIEW

File: sway-lsp/src/utils/test.rs [hunk 1]
```diff
@@ -16,8 +20,8 @@ pub(crate) fn e2e_test_dir() -> PathBuf {
         .join("struct_field_access")
 }
 
-pub(crate) fn sway_example_dir() -> PathBuf {
```
Comment (sdankel) [inline sway-lsp/src/utils/test.rs:19]: This wasn't being used

File: sway-lsp/src/core/token_map.rs [hunk 2]
```diff
@@ -22,18 +22,16 @@ impl TokenMap {
         &'s self,
         uri: &'s Url,
     ) -> impl 's + Iterator<Item = (Ident, Token)> {
-        self.iter()
-            .filter(|item| {
-                let (_, span) = item.key();
-                match span.path() {
-                    Some(path) => path.to_str() == Some(uri.path()),
-                    None => false,
+        self.iter().flat_map(|item| {
```
Comment (sdankel) [inline sway-lsp/src/core/token_map.rs:25]: Slightly easier to read with flat_map, imho

**Review by mohammadfawaz — APPROVED** [review]
Changes in `sway-core` seem reasonable to me! This does indeed the behavior of `rustc` which I really like.

File: sway-lsp/test/fixtures/diagnostics/dead_code/Forc.toml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+[project]
+authors = ["Sophie"]
```
Comment (otrho) [inline sway-lsp/test/fixtures/diagnostics/dead_code/Forc.toml:2]: Just noticed this.  We use `authors = ["Fuel Labs <contact@fuel.sh>"]` in the compiler tests, which is automatic if you set `FORC_INIT_MANIFEST_AUTHOR='Fuel Labs <contact@fuel.sh>'` in your env.

John used to be really strict with it but perhaps it isn't so important these days.

**Review by otrho — DISMISSED** [review]

File: sway-lsp/test/fixtures/diagnostics/dead_code/Forc.toml (hunk 3 above)
Comment (mohammadfawaz) [reply sway-lsp/test/fixtures/diagnostics/dead_code/Forc.toml:2]: Actually yeah it's probably better to be consistent and use `Fuel Labs`.

File: sway-lsp/src/utils/test.rs (hunk 1 above)
Comment (JoshuaBatty) [reply sway-lsp/src/utils/test.rs:19]: Can we put this function back in? Although it's not being used in any current CI tests, it is handy to have when testing locally.

File: sway-lsp/test/fixtures/diagnostics/dead_code/Forc.toml (hunk 3 above)
Comment (JoshuaBatty) [reply sway-lsp/test/fixtures/diagnostics/dead_code/Forc.toml:2]: Agree with the above.

**Review by JoshuaBatty — CHANGES_REQUESTED** [review]
This is really great work @sdankel! just one tiny nit and I'm happy to approve.

**Review by mohammadfawaz — APPROVED** [review]

**Review by JoshuaBatty — APPROVED** [review]
:shipit:

**Pull request merged**